### ISSUE TITLE
Reorganizing: Remove existing regions affects in-method regions as well

### ISSUE THREAD

**Issue by khorvat** [codecadwallader/codemaid#272]
### Environment
- Visual Studio version: 2013 Premium Update 5
- CodeMaid version: 10.0.72
- Code language: C#

### Description
Reorganizing not working properly in case when you add new methods or properties to existing class.

### Steps to recreate
* Insert this code

`
    using System;
    using System.Web.Mvc;

    namespace Namespace
    {
        /// <summary>
        /// Login controller.
        /// </summary>
        public class Login2Controller : Controller
        {
            protected IMembershipService Service { get; private set; }

            public ActionResult Index2()
            {
                return View();
            }

            #region Constructors

            /// <summary>
            /// Initializes a new instance of the <see cref="Login2Controller" /> class.
            /// </summary>
            /// <param name="service">The service.</param>
            public Login2Controller(IMembershipService service)
            {
                Service = service;
            }

            #endregion Constructors

            #region Methods

            /// <summary>
            /// Index view.
            /// </summary>
            /// <returns></returns>
            public ActionResult Index()
            {
                return View();
            }

            #endregion Methods
        }
    }
`
* try to reorganize
* my settings defines region order as follows: ctor, props, methods
* review order based on my settings you will notice some strange 

### Current behavior
If constructor is in region, property is just surrounded with region and left in place - wrong place 

### Expected behavior
Respect defined settings and place property in the right order then surround with region.

**Comment by codecadwallader** [comment]
Thanks for reporting the issue.  I suspect you have enabled "Insert new regions" but you have not enabled "Remove existing regions".

Items that are inside existing regions are ignored at reorganization time (unless you have also disabled the "Keep members within regions" option).  Regions could contain a mix of different items (e.g. constructors and methods) so we don't give them any sort priority and they are left in place.

If you enable the "Remove existing regions" setting as well, then CodeMaid will be able to fully sort the file as it sounds like you would desire.

**Comment by khorvat** [comment]
I have enabled "Remove ..." and tried to enable/disable "Keep ...", now members are placed in appropriate regions but regions are not sorted as per order defined in the settings. Do you have any other suggestions on how to amend settings to get it working ?

Thanks

**Comment by khorvat** [comment]
One more thing, since I have enabled "Remove ..." regions inside the methods are removed, is that a desired behavior ?

**Comment by codecadwallader** [comment]
Can you please attach an export of your settings?

**Comment by khorvat** [comment]
Here you are, thanks.
[CodeMaid.zip](https://github.com/codecadwallader/codemaid/files/216573/CodeMaid.zip)

**Comment by codecadwallader** [comment]
Thanks for the export, but it doesn't contain any of the settings you referenced above.  I suspect you exported your solution-specific settings, but those settings may be defined at the user settings level.  Can you export that as well?

**Comment by khorvat** [comment]
Sorry about that, here you are.

[CodeMaid.zip](https://github.com/codecadwallader/codemaid/files/217199/CodeMaid.zip)
![screenshot_1](https://cloud.githubusercontent.com/assets/1308966/14495204/c18e0df8-018e-11e6-8f66-53aa95f34fe6.png)

**Comment by codecadwallader** [comment]
Thanks for your settings.  It looks like you have the setting "Primary ordering should be by access then type" specified.  That's why the public methods are appearing above the protected property.  If you reset that to the default "type then access" then I think you'll see the behavior you expect where type takes precedence.

![image](https://cloud.githubusercontent.com/assets/3823772/14504345/681e5dc4-0182-11e6-8699-9831cb9c1f85.png)

Hope it helps.

**Comment by khorvat** [comment]
I'm referring to region order not member order. I'm fine with the setting access then type, my problem right now is that properties are pushed to bottom of the class bellow methods. It is not the order I have in settings and now all my classes, from git perspective, looks completely changes and other devs have trouble seeing the diffs.

Thanks

**Comment by codecadwallader** [comment]
I'm confused.. if you have the ordering set to "access then type".. then your public methods are supposed to appear above your protected properties.  If you want to see your type settings (e.g. properties before methods) take precedence, then you need to reset primary ordering to be by type then access.  Regions can be taken out of the equation, they'll simply decorate wherever the members are ordered.

**Comment by khorvat** [comment]
Well in 0.9 version members were ordered inside the region and regions were ordered as per order in settings?

**Comment by codecadwallader** [comment]
The behavior should not have changed since v0.9.. regions have never been sorted, only their contents.  Why not change the setting back to "type then access" since that's the behavior you want?

**Comment by khorvat** [comment]
Sorry about misunderstanding about regions, I was tracking behavior with regions so I mentioned regions but was referring to types. Ok I noticed that I had "access then type" and that was wrong for me, after I have reverted to my old settings few things are working as expected but few still don't.

1. I have "Remove regions ..." turned on, I noticed that the regions inside the methods are removed, is that ok ?
2. We noticed some difference using 0.9 and 10.0 on two different machines using same settings, "Remove regions - on", "Type then access - on", "Insert regions - on" - using these settings will push properties to the bottom outside of any region on 0.9 and on 10.0 will do it right

Any ideas ?

**Comment by codecadwallader** [comment]
1.  Correct, we don't gracefully handle regions inside methods during reorganization.  That can be a separate bug report (or a re-purposing of this one).
2.  Do you have a code sample?  Or if it's working as you expect it to now.. does it need any further investigation?

**Comment by khorvat** [comment]
1. I'll open a new bug r.
2. Just wanted to let you know there is a diff. Now it's working fine beside the above

**Comment by codecadwallader** [comment]
Let's just re-purpose this one since there's only one issue:

During reorganizing, if remove existing regions is enabled (off by default) then all regions are removed.. even those that are within methods and shouldn't really participate in reorganization.

Before:
```csharp
namespace ClassLibrary2
{
    internal class Issue272
    {

        #region Public Methods

        public int AMethod()
        {
            #region Declarations

            int x = 5;

            #endregion Declarations

            #region Calculations

            int y = x * 2;

            #endregion Calculations

            #region Returns

            return y;

            #endregion Returns
        }

        #endregion Public Methods
    }
}
```

Observed
```csharp
namespace ClassLibrary2
{
    internal class Issue272
    {
        #region Methods

        public int AMethod()
        {
            int x = 5;

            int y = x * 2;

            return y;
        }

        #endregion Methods
    }
}
```

**Comment by codecadwallader** [comment]
I have a proposed fix for this so that regions inside methods will be ignored.  This will be available momentarily in our CI channel (v10.0.84 or higher) if you would like to test it: http://vsixgallery.com/extension/4c82e17d-927e-42d2-8460-b473ac7df316/

**Comment by khorvat** [comment]
I can confirm that this is working correctly now, method regions are not removed anymore. Sorry for a late answer.

**Comment by codecadwallader** [comment]
Cool, thanks for confirming!

**Comment by khorvat** [comment]
On more thing, what about regions inside regions ? I saw a corner case in my code where inner regions are removed too ?

**Comment by codecadwallader** [comment]
We had a fix back in v0.9.1 (see #213).. if you are still seeing an active issue can you please open a new issue?

**Comment by khorvat** [comment]
Ok, I'll open another issue for this.

**Comment by codecadwallader** [comment]
Sounds good.  I'm going to reopen this issue until this issue is officially released in v10.1 so it is more discoverable to anyone else who may be encountering it currently.
